lib/redmine_s3/attachment_patch.rb

```diff
@@ -30,12 +31,44 @@ def put_to_s3
       end
 
       def delete_from_s3
-        logger.debug("Deleting #{disk_filename}")
-        RedmineS3::Connection.delete(disk_filename)
+        logger.debug("Deleting #{disk_filename_s3}")
+        RedmineS3::Connection.delete(disk_filename_s3)
       end
 
       # Prevent file uploading to the file system to avoid change file name
       def files_to_final_location; end
+
+      # Returns the full path the attachment thumbnail, or nil
+      # if the thumbnail cannot be generated.
+      def thumbnail_s3(options={})
+        return nil unless Object.const_defined?(:Magick)
+        if thumbnailable?
+          size = options[:size].to_i
+          if size > 0
+            # Limit the number of thumbnails per image
+            size = (size / 50) * 50
+            # Maximum thumbnail size
+            size = 800 if size > 800
+          else
+            size = Setting.thumbnails_size.to_i
+          end
+          size         = 100 unless size > 0
+          target       = "#{id}_#{digest}_#{size}.thumb"
+          update_thumb = options[:update_thumb] || false
+          begin
+            RedmineS3::ThumbnailPatch.generate_s3_thumb(self.disk_filename_s3, target, size, update_thumb)
+          rescue => e
+            logger.error "An error occured while generating thumbnail for #{disk_filename_s3} to #{target}\nException was: #{e.message}" if logger
+            return nil
+          end
+        end
+      end
+
+      def disk_filename_s3
+        path = disk_filename
+        path = File.join(disk_directory, path) unless disk_directory.blank?
+        path
```

Comment:
Your example will return nil because the entire line will not be executed if `disk_directory.blank?` is true. And my code returns `disk_filename` in this case. Well there's probably no need for the last line.